base/src/main/scala/typeclass/MonadInstancesStdLib.scala

```diff
@@ -3,15 +3,15 @@ package typeclass
 
 trait MonadInstancesStdLib {
   implicit val option: Monad[Option] = new MonadClass[Option] {
-    override def ap[A, B](oa: Option[A])(f: Option[A => B]): Option[B] = oa.flatMap(a => f.map(_(a)))
-    override def flatMap[A, B](oa: Option[A])(f: A => Option[B]): Option[B] = oa.flatMap(f)
+    override def ap[A, B](oa: => Option[A])(f: Option[A => B]): Option[B] = oa.flatMap(a => f.map(_(a)))
+    override def flatMap[A, B](oa: => Option[A])(f: A => Option[B]): Option[B] = oa.flatMap(f)
     override def map[A, B](oa: Option[A])(f: A => B): Option[B] = oa.map(f)
     override def pure[A](a: A): Option[A] = Option(a)
   }
 
   implicit val list: Monad[List] = new MonadClass[List] {
-    override def ap[A, B](xs: List[A])(f: List[A => B]): List[B] = xs.flatMap(a => f.map(_(a)))
-    override def flatMap[A, B](xs: List[A])(f: A => List[B]): List[B] = xs.flatMap(f)
+    override def ap[A, B](xs: => List[A])(f: List[A => B]): List[B] = xs.flatMap(a => f.map(_(a)))
```

Comment:
This is backwards wrt the default definition in `BindClass`; the `f` effect should be "first".